Rollback parent transaction when children fails to update

gmalette opened the pull request:
## Summary

Rails supports autosave associations on the owner of a `has_many`
relationship. In certain situation, if the children of the association
fail to save, the parent is not rolled back. This PR starts making it rollback.

## Details

```ruby
class Employee < ActiveRecord::Base
end

class Company < ActiveRecord::Base
  has_many(:employees)
end

company = Company.new
employee = company.employees.new
company.save
```

In the previous example, if the Employee failed to save, the Company
will not be rolled back. It will remain in the database with no
associated Employee.

I expect the `company.save` call to be atomic, and either create all or
none of the records.

The persistence of the Company already starts a transaction that nests
it's associated record updates. However, it didn't track the success or failure of it's
associations in this very situation, and the outermost transaction is not
rolled back.

This PR makes the change to track the success of the child insertion and
rollback the parent if any of the children fail.

gmalette committed:
Rollback parent transaction when children fails to update

Rails supports autosave associations on the owner of a `has_many`
relationship. In certain situation, if the children of the association
fail to save, the parent is not rolled back.

```ruby
class Employee < ActiveRecord::Base
end

class Company < ActiveRecord::Base
  has_many(:employees)
end

company = Company.new
employee = company.employees.new
company.save
```

In the previous example, if the Employee failed to save, the Company
will not be rolled back. It will remain in the database with no
associated Employee.

I expect the `company.save` call to be atomic, and either create all or
none of the records.

The persistance of the Company already starts a transaction that nests
it's children. However, it didn't track the success or failure of it's
children in this very situation, and the outermost transaction is not
rolled back.

This PR makes the change to track the success of the child insertion and
rollback the parent if any of the children fail.

gmalette commented on activerecord/lib/active_record/autosave_association.rb:
```diff
@@ -401,7 +401,10 @@ def save_collection_association(reflection)
                 else
-                  association.insert_record(record) unless reflection.nested?
+                  association_saved = association.insert_record(record) unless reflection.nested?
+                  if reflection.options.fetch(:validate, true)
```
Is `validate: nil` considered true?

eugeneius replied:
It looks like it is:

https://github.com/rails/rails/blob/df850b8d7126044a820291bbfd2bce80636a442c/activerecord/lib/active_record/autosave_association.rb#L25-L27

We should be calling `reflection.validate?` here though, as `define_autosave_validation_callbacks` does:

https://github.com/rails/rails/blob/df850b8d7126044a820291bbfd2bce80636a442c/activerecord/lib/active_record/autosave_association.rb#L213

gmalette commented on activerecord/test/cases/autosave_association_test.rb:
```diff
@@ -517,6 +517,24 @@ def test_invalid_adding_before_save
+    assert_predicate client, :valid?
+    assert_predicate new_firm, :valid?
+    assert_not new_firm.save
+    assert new_firm, :persisted?
```
I'm not happy that `new_firm` is `persisted?`, but this is a side-effect of the insert of `new_firm` succeeding and the database providing an ID, even if the transaction is rolled back later. Thoughts?

Edouard-chin replied:
I think this might be prone to obscure bug inside application

> new_firm.persisted? => true
new_firm.new_record? => false
Firm.exists?(new_firm.id) => false

Should we instead reset the `@new_record` ivar when the transaction is rolled back?

gmalette replied:
If everyone agrees that's the solution, I'm all for it and I can do it, I just didn't want to change the existing behaviour

gmalette replied:
What about the ID and other database-provided attributes?

gmalette replied:
More on this: it looks like there is already code that handles rolling back the attributes

https://github.com/rails/rails/blob/8ae38dab5251613a7e7f5af5289e524abca20efd/activerecord/lib/active_record/transactions.rb#L323-L327

However, it looks like `ActiveRecord::Rollback` is swallowed before it reaches that point. 

https://github.com/rails/rails/blob/8ae38dab5251613a7e7f5af5289e524abca20efd/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L247-L257

I'm absolutely not familiar with this code so I'd really appreciate pointers

eugeneius replied:
https://github.com/rails/rails/pull/32862 should fix the rollback behaviour so that `new_firm.persisted?` is false.

Incidentally, the `assert` here will never fail, because `new_firm` will always be truthy; it looks like it should be `assert_predicate` instead.

gmalette replied:
> #32862 should fix the rollback behaviour so that new_firm.persisted? is false.

Awesome! I'll rebase and make the changes when it gets merged.

> Incidentally, the assert here will never fail

😂 that takes me back to #20527

eugeneius replied:
😄 

I've also bumped into this problem before: https://github.com/cypriss/mutations/pull/110

gmalette committed:
Change the test to reflect what we expect

Once #32862 is merged, rolling back a record will rollback it's state to match
the state before the database changes were applied

gmalette commented:
Now that #32862 is merged I rebased, and the tests are passing

eugeneius commented on activerecord/lib/active_record/autosave_association.rb:
```diff
@@ -401,7 +401,10 @@ def save_collection_association(reflection)
                   saved = association.insert_record(record, false)
                 else
-                  association.insert_record(record) unless reflection.nested?
+                  association_saved = association.insert_record(record) unless reflection.nested?
```
If `reflection.nested?` is true, we should probably skip reassigning `saved`.

eugeneius commented on activerecord/test/cases/autosave_association_test.rb:
```diff
@@ -517,6 +517,24 @@ def test_invalid_adding_before_save
+  def test_adding_unsavable_association
+    new_firm = Firm.new("name" => "A New Firm, Inc")
+    client = new_firm.clients.new("name" => "Apple")
+    # Stub the save method of the client to simulate aborting the insert
```
Is this simulating `throw :abort` in a before callback? Could we make that happen for real? Rails' tests are generally written end-to-end, using only the public API.

gmalette replied:
I actually came across the bug addressed in this PR trying to fix https://github.com/Shopify/activerecord-rescue_from_duplicate/issues/13, and I just replicated it's [behaviour](https://github.com/Shopify/activerecord-rescue_from_duplicate/blob/ca140ad0102db088ba671573d9490e96d92c344c/lib/rescue_from_duplicate/active_record/extension.rb#L29), but I think it will work with `throw :abort` in a `before_save` just as well.

eugeneius commented on activerecord/test/cases/autosave_association_test.rb:
```diff
@@ -517,6 +517,24 @@ def test_invalid_adding_before_save
+    assert_predicate new_firm, :valid?
+    assert_not new_firm.save
+    assert_not_predicate new_firm, :persisted?
+    assert_not Firm.exists?(new_firm.id)
```
Now that `persisted?` works correctly, this feels redundant.

gmalette committed:
Use only the public API to express the tests

gmalette committed:
Refactor to avoid reassigning saved for nested reflections

gmalette commented:
@eugeneius I refactored to use the public API with `throw :abort`, along with your other suggestions

eugeneius reviewed (approved):
This now looks good to me! 👍

rafaelfranca commented:
Thanks!

ghiculescu commented:
Hey @gmalette 	just wondering if you would consider https://github.com/rails/rails/issues/35009 to be a regression?

gmalette commented:
I’m not familiar enough with the contract of AR with regards to `validate: false` to know for sure, but to me it looks like a regression
